src/envs/pddl_procedural_generation.py

```diff
@@ -186,3 +191,262 @@ def _generate_delivery_problem(num_locs: int, num_want_locs: int,
+    for i, path_coord in enumerate(random_path):
+        loc_prob = rng.uniform()
+        if path_coord == (I_row, I_col):
+            grid[path_coord[0]][path_coord[1]] = FOREST_I
+        elif path_coord == (G_row, G_col):
+            grid[path_coord[0]][path_coord[1]] = FOREST_G
+        elif i > 1 and not last_was_hill and loc_prob <= 0.2:
+            grid[path_coord[0]][path_coord[1]] = FOREST_H
+            last_was_hill = True
+        else:
+            grid[path_coord[0]][path_coord[1]] = FOREST_P
+
+    for r in range(grid_height):
+        for c in range(grid_width):
+            assert grid[r][c] != -1
+
+    return grid
+
+
+def _random_grid_walk(
+        curr_coords: Tuple[int, int], goal_coords: Tuple[int, int],
+        visited: Set[Tuple[int, int]], grid_height: int, grid_width: int,
+        previous_coords: Optional[Tuple[int, int]],
+        rng: np.random.Generator) -> Optional[List[Tuple[int, int]]]:
+    if curr_coords == goal_coords:
+        return [curr_coords]
+
+    for delta in rng.permutation([[0, 1], [1, 0], [0, -1], [-1, 0]]):
+        new_coord = (curr_coords[0] + delta[0], curr_coords[1] + delta[1])
+        # Out of bounds.
+        if new_coord[0] < 0 or new_coord[0] >= grid_height or new_coord[
+                1] < 0 or new_coord[1] >= grid_width:
+            continue
+
+        # Already visited.
+        if new_coord in visited:
+            continue
+
+        # Prevent visiting coords that are adjacent to visited coords, except
+        # for the most recent predecessor.
+        adjacent_excluding_previous = {
+            (curr_coords[0] + adj_delta[0], curr_coords[1] + adj_delta[1])
+            for adj_delta in [[0, 1], [1, 0], [0, -1], [-1, 0]]
+        } - {previous_coords}
+        adjacent_hit = False
+        for adjacent_coord in adjacent_excluding_previous:
+            if adjacent_coord in visited:
+                adjacent_hit = True
+        if adjacent_hit:
+            continue
+
+        # Prevent visiting unreachable coordinates.
+        if not _random_walk_reachable(new_coord, goal_coords,
+                                      visited | {curr_coords}, grid_height,
+                                      grid_width):
+            continue
+
+        # Successfully extended the path.
+        grid_walk_from_child = _random_grid_walk(new_coord, goal_coords,
+                                                 visited | {curr_coords},
+                                                 grid_height, grid_width,
+                                                 curr_coords, rng)
+        if grid_walk_from_child is not None:
+            return [curr_coords] + grid_walk_from_child
+
+    return None
+
+
+def _random_walk_reachable(curr_coords: Tuple[int,
+                                              int], goal_coords: Tuple[int,
+                                                                       int],
+                           prev_visited: Set[Tuple[int, int]],
```

Comment:
this function seems rather similar to the above and I'm struggling to understand how it's being used. Maybe add a comment here or above to explain what's happening just so people editing in the future can grok this easily?